### ISSUE TITLE
Typing issues with interfaces and implementations

### ISSUE THREAD

**Issue by mtraynham** [500tech/angular-tree-component#497]
I'm getting this TypeScript exception

```
(5,22): Class 'TreeModel' incorrectly implements interface 'ITreeModel'.
  Types of property 'options' are incompatible.
    Type 'TreeOptions' is not assignable to type 'ITreeOptions'.
      Types of property 'allowDrag' are incompatible.
        Type '(node: TreeNode) => boolean' is not assignable to type 'boolean | IAllowDragFn'.
          Type '(node: TreeNode) => boolean' is not assignable to type 'IAllowDragFn'.
            Types of parameters 'node' and 'node' are incompatible.
              Type 'ITreeNode' is not assignable to type 'TreeNode'.
                Property 'treeModel' is missing in type 'ITreeNode'.
```

It also seems weird that I'm importing `ITreeNode` from `import {ITreeNode} from 'angular-tree-component/dist/defs/api';` because it's not part of the lib export, even though the ITreeOptions explicitly depends on it.

If I were to use `TreeNode`, say in `ITreeOptions.allowDrag`, `allowDrag: (node: TreeNode): boolean => !node.isRoot`, I get a TypeScript error:
```
Error:(67, 9) TS2322: Type '{ idField: string; displayField: string; allowDrag: (node: TreeNode) => boolean; allowDrop: (_e: ...' is not assignable to type 'ITreeOptions'.
  Types of property 'allowDrag' are incompatible.
    Type '(node: TreeNode) => boolean' is not assignable to type 'boolean | IAllowDragFn'.
      Type '(node: TreeNode) => boolean' is not assignable to type 'IAllowDragFn'.
        Types of parameters 'node' and 'node' are incompatible.
          Type 'ITreeNode' is not assignable to type 'TreeNode'.
            Property 'treeModel' is missing in type 'ITreeNode'.
```

**Comment by adamkleingit** [comment]
What exactly are you doing? Can you reproduce in plunkr?
https://plnkr.co/edit/u5fMmaM14rJ0aqG3k240?p=preview

Actually you shouldn't use ITreeNode, it is for docs purpose only


*Adam Klein*
*Chief Technology Officer*



*Tel.*   +972-52-7475633 <+972527475633>
*Web*  500tech.com <http://www.500tech.com/> | github.com/500tech |
@adamklein500 <http://twitter.com/adamklein500>

**Comment by mtraynham** [comment]
Hey @adamkleingit 

It seems like two main issues, so I'll separate them out. This is TypeScript 2.6.1

### 1. Importing `ITreeOptions` gives me a TypeScript error for even simple usage:
```ts
import {ITreeOptions} from 'angular-tree-component';

class FooBar {
    public treeOptions: ITreeOptions;
    constructor () {
        this.treeOptions = {
            idField: 'id',
            displayField: 'data.name'
        };
    }
}

export default FooBar;
```
has TypeScript error:
```ts
node_modules/angular-tree-component/dist/models/tree.model.d.ts
Error:(5, 22) TS2420: Class 'TreeModel' incorrectly implements interface 'ITreeModel'.
  Types of property 'options' are incompatible.
    Type 'TreeOptions' is not assignable to type 'ITreeOptions'.
      Types of property 'allowDrag' are incompatible.
        Type '(node: TreeNode) => boolean' is not assignable to type 'boolean | IAllowDragFn'.
          Type '(node: TreeNode) => boolean' is not assignable to type 'IAllowDragFn'.
            Types of parameters 'node' and 'node' are incompatible.
              Type 'ITreeNode' is not assignable to type 'TreeNode'.
                Property 'treeModel' is missing in type 'ITreeNode'.
```

### 2. Using `TreeNode` with `ITreeOptions` throws an error, so instead use `ITreeNode`
```ts
import {ITreeOptions, TreeNode} from 'angular-tree-component';

class FooBar {
    public treeOptions: ITreeOptions;
    constructor () {
        this.treeOptions = {
            idField: 'id',
            displayField: 'data.name',
            allowDrag: (node: TreeNode): boolean => !node.isRoot
        };
    }
}

export default FooBar;
```
Throws a second TypeScript error on top of the first one from above
```sh
src/app/foo/views/foobar.component.ts
Error:(6, 9) TS2322: Type '{ idField: string; displayField: string; allowDrag: (node: TreeNode) => boolean; }' is not assignable to type 'ITreeOptions'.
  Types of property 'allowDrag' are incompatible.
    Type '(node: TreeNode) => boolean' is not assignable to type 'boolean | IAllowDragFn'.
      Type '(node: TreeNode) => boolean' is not assignable to type 'IAllowDragFn'.
        Types of parameters 'node' and 'node' are incompatible.
          Type 'ITreeNode' is not assignable to type 'TreeNode'.
            Property 'treeModel' is missing in type 'ITreeNode'.
```

#### Importing `ITreeNode` directly works, but it's not exported from root typing
This removes the second error, but the first still shows.
```ts
import {ITreeOptions} from 'angular-tree-component';
import {ITreeNode} from 'angular-tree-component/dist/defs/api';

class FooBar {
    public treeOptions: ITreeOptions;
    constructor () {
        this.treeOptions = {
            idField: 'id',
            displayField: 'data.name',
            allowDrag: (node: ITreeNode): boolean => !node.isRoot
        };
    }
}

export default FooBar;
```

**Comment by m-ghaoui** [comment]
I got these kinds of errors when I set `"strict": true` in the `tsconfig.json` file and running `ng serve`

```
ERROR in node_modules/angular-tree-component/dist/models/tree.model.d.ts(5,22): error TS2420: Class 'TreeModel' incorrectly implements interface 'ITreeModel'.
  Types of property 'options' are incompatible.
    Type 'TreeOptions' is not assignable to type 'ITreeOptions'.
      Types of property 'allowDrag' are incompatible.
        Type '(node: TreeNode) => boolean' is not assignable to type 'boolean | IAllowDragFn'.
          Type '(node: TreeNode) => boolean' is not assignable to type 'IAllowDragFn'.
            Types of parameters 'node' and 'node' are incompatible.
              Type 'ITreeNode' is not assignable to type 'TreeNode'.
                Property 'treeModel' is missing in type 'ITreeNode'.
```

**Comment by adamkleingit** [comment]
I can't reproduce, can you send link to repo?

**Comment by m-ghaoui** [comment]
Sorry it's an internal project which I may not publish.

But I think you can easily reproduce it by creating a new Angular 5 project with `ng new` and then setting this in your tsconfig.json:

```
{
  "compileOnSave": false,
  "compilerOptions": {
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "strict": true,
    "noStrictGenericChecks": true,
    "strictNullChecks": true,
    "target": "es5",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2017",
      "dom"
    ]
  },
  "angularCompilerOptions": {
    "preserveWhitespaces": false
  },
  "include": [
    "src/**/*"
  ]
}
```

Note I'm setting `strict` and `strictNullChecks` to `true`.

**Comment by adamkleingit** [comment]
OK I found the problem. I was using typescript 2.4 and the problem didn't happen there. I'll see if this is a typescript problem or tree problem

**Comment by adamkleingit** [comment]
Solved in 7.0.1
